File: Lab4.cpp
//UWYO COSC 2030
//Lab 4
//Written by Danny Radosevich
//8-4-19

#include <iostream>
#include <stack>
#include <queue>
#include <vector>

using namespace std;


//function declarations
string stringReverse(string toReverse);
bool parenCheck(string toCheck);

/*DO NOT CHANGE MAIN OR THE FUNCTION DECLARATIONS*/
int main()
{
  //paren strings
  string paren1 = "(()((((()()()((()(((()()()()(((()(()()()(())()())()()))))()()()))()))()())())())))";
  string paren2 = "()((((((()(((((()((()()()()()(()))))))))()))()(())(())(((((()()(())))(()()())(()))";
  string paren3 = "(()((()(()()(()(((((()()(()()()((((()()(()()))()))))()()())))))()()())()()))())())";
  string paren4 = "(()()(((()()(()(()()(()()()()()()(()(((((((((())())))))()))))()()))()())()()))()))";
  string paren5 = "((())";
  string paren6 = ")))(((";

  string rev1 = "sekopog";
  string rev2 = "racecar";
  string rev3 = "regnolsignirtssiht";
  string rev4 = "wonybnwodsihtevahyllufepohdluohssyuguoy";

  //Ignore this for now, these vectors are just part of the program
  vector<string> parens;
  parens.push_back(paren1);
  parens.push_back(paren2);
  parens.push_back(paren3);
  parens.push_back(paren4);
  parens.push_back(paren5);
  parens.push_back(paren6);

  vector<string> rever;
  rever.push_back(rev1);
  rever.push_back(rev2);
  rever.push_back(rev3);
  rever.push_back(rev4);
  
  //call the parenCheck
  for(int i=0; i<6; i++)
  {
    cout<<"Checking string"<<i+1<<endl;
    if(parenCheck(parens.at(i)))
    {
      cout<<"Your code says that string"<<i+1<<" is good.";
      if(i == 0 || i == 2 || i==3)
      {
        cout<<" Correct!"<<endl<<endl;
      }
      else
      {
        cout<<" Incorrect!"<<endl<<endl;
      }
    }
    else
    {
      cout<<"Your code says that string"<<i+1<<" is bad.";
      if(i == 1 || i == 4 || i==5)
      {
        cout<<" Correct!"<<endl<<endl;
      }
      else
      {
        cout<<" Incorrect!"<<endl<<endl;
      }
    }
  }

  //now to reverse some strings
  for(int i =0; i<4; i++)
  {
    cout<<"Reversing string"<<i+1<<endl;
    cout<<stringReverse(rever.at(i))<<endl<<endl;
  }
}

//This function takes a string and reverses it using a stack
//it returns the reversed string
string stringReverse(string toReverse)
{
  stack<char> reverse; //declare the stack
  string newString = "";

  // Put every letter into a stack, so it may be read in reverse
  for (char letter : newString)
  {
    reverse.push(letter);
  }

  // Move trhough stack to create string
  while (!reverse.empty())
  {
    // Pull item from stack
    char letter = reverse.top();
    reverse.pop();

    // Add to string
    newString = (newString + letter);
  }

  return newString;
}

//this function uses a queue to check if a string of parens is properly matched
bool parenCheck(string toCheck)
{
  queue<char> checker; //declare my queue

  // We can check if each parenthesis is matched by adding a token to the queue once finding an open parenthesis,
  // and removing a token when meeting a closed parenthesis
  for (char letter : toCheck)
  {
    if (letter == '(')
    {
        // Open a Parenthesis
        checker.push(letter);
    }
    else if (letter == ')' && !checker.empty())
    {
        // Close a parenthesis
        checker.pop();
    }
  }

  // Now we check if the queue has successfully been cleared
  if (checker.empty())
  {
    return true;
  }

  return false; //need a return statement to compile
}

File: Lab4.py
# Russell Todd
# Lab 4 for UWYO COSC 2030
# Stacks and Queues
#
# You'll be checking parentheses and reversing strings just like you did in C++
# However, in Python the class you'll be using is a List

paren1 = "(()((((()()()((()(((()()()()(((()(()()()(())()())()()))))()()()))()))()())())())))";
paren2 = "()((((((()(((((()((()()()()()(()))))))))()))()(())(())(((((()()(())))(()()())(()))";
paren3 = "(()((()(()()(()(((((()()(()()()((((()()(()()))()))))()()())))))()()())()()))())())";
paren4 = "(()()(((()()(()(()()(()()()()()()(()(((((((((())())))))()))))()()))()())()()))()))";
paren5 = "((())";
paren6 = ")))(((";

rev1 = "sekopog";
rev2 = "racecar";
rev3 = "regnolsignirtssiht";
rev4 = "wonybnwodsihtevahyllufepohdluohssyuguoy";

parens = [paren1, paren2, paren3, paren4, paren5, paren6]
rever = [rev1, rev2, rev3, rev4]

# Finish this function
# You must complete it using a List in a similar manner as a Queue
def parenCheck(toCheck):
    pseudoQueue = []
    
    # Check string for open and closed parenthesis
    # If the queue is empty, it means all parenthesis have been matched.
    
    for letter in toCheck:
        # Open Parens
        if letter == "(":
            pseudoQueue.append(letter)
            
        # Closed Parens
        elif (letter == ")" and len(pseudoQueue) != 0):
            pseudoQueue.pop()
            
    # Check queue length
    if len(pseudoQueue) == 0:
        return True
    else:
        return False

# Finish this function
# You must complete it using a List in a similar manner as a Stack
def stringReverse(toReverse):
    pseudoStack = []
    newString = ""

    # Add letters to stack
    for letter in toReverse:
        pseudoStack.insert(0, letter)
        
    # Remove letters from stack
    for item in pseudoStack:
        newString = newString + item
        
    return newString

# Time to check some parentheses
for i in range(6):
    print("\nChecking string"+str(i+1))
    if parenCheck(parens[i]):
        print("Your code says that string"+str(i+1)+" is good. ")
        if i == 0 or i == 2 or i == 3:
            print("Correct!")
        else:
            print("Incorrect!")
    else:
        print("Your code says that string"+str(i+1)+" is bad. ")
        if i == 1 or i == 4 or i == 5:
            print("Correct!")
        else:
            print("Incorrect!")

# Time to reverse some strings
for j in range(4):
    print("\nReversing string"+str(j+1))
    print(stringReverse(rever[j]))
